src/pf/existing/update_source_files.cpp

```diff
@@ -0,0 +1,190 @@
+                    std::vector<std::string> const& sources,
+                    std::string const&              indent) {
+    bool first_iteration = true;
+
+    for (auto const& source : sources) {
+        if (!first_iteration) {
+            insertion_point = std::next(cmakelists.insert(insertion_point, '\n'));
+        } else {
+            first_iteration = false;
+        }
+
+        insertion_point
+            = std::next(cmakelists.insert(insertion_point, indent.begin(), indent.end()),
+                        indent.size());
+        insertion_point
+            = std::next(cmakelists.insert(insertion_point, source.begin(), source.end()),
+                        source.size());
+    }
+
+    return insertion_point;
+}
+
+std::pair<std::string::iterator, std::string::iterator>
+write_sources(std::string&                    cmakelists,
+              std::vector<std::string> const& source_strings,
+              std::string::iterator           insertion_comment,
+              std::string::iterator           begin_fn,
+              std::string::iterator           end_fn) {
+    std::string const indent
+        = ::get_indent_at_indicator_comment(insertion_comment, begin_fn, end_fn);
+    auto insertion_point = ::find_insertion_location(insertion_comment, begin_fn, end_fn);
+
+    // Note: invalidates other iterators
+    insertion_point = ::erase_existing_sources(cmakelists, insertion_point, begin_fn, end_fn);
+
+    auto const end_insertion
+        = ::insert_sources(cmakelists, insertion_point, source_strings, indent);
+    return std::pair{end_insertion, cmakelists.end()};
+}
+
+}  // namespace
+
+void pf::update_source_files(fs::path const&              project_root,
+                             std::vector<fs::path> const& source_files) {
+    fs::path const src_dir        = project_root / "src";
+    fs::path const src_cmakelists = src_dir / "CMakeLists.txt";
+
+    if (!fs::exists(src_cmakelists)) {
+        throw std::system_error{
+            std::make_error_code(std::errc::no_such_file_or_directory),
+            src_cmakelists.string() + " does not exist",
+        };
+    }
+
+    std::string cmakelists = pf::slurp_file(src_cmakelists);
+
+    std::vector<std::string> const source_strings
+        = ::relative_source_strings(source_files, src_dir);
+
+    std::string const cmakelists_cpy = cmakelists;
+
+    auto begin = cmakelists.begin();
+    auto end   = cmakelists.end();
+
+    while (begin != end) {
+        auto const begin_fn          = ::find_next_function(begin, end);
+        auto const end_fn            = ::find_end_function(begin_fn, end);
+        auto const insertion_comment = ::find_insertion_indicator_comment(begin_fn, end_fn);
+
+        if (insertion_comment == end_fn) {
+            begin = end_fn;
+            continue;
+        }
+
+        std::tie(begin, end)
+            = ::write_sources(cmakelists, source_strings, insertion_comment, begin_fn, end_fn);
```

Comment:
Don't allow multiple `# sources` to expand this. Instead, error out if we see multiple source-list insertion points.

Why? If the user needs to do this, they can put it in a `set(...)`. Requiring exactly one `# sources` indicator means that heuristics could be used to find the source list, not just an indicator comment.

Reply:
If we just say, "this feature is experimental" or say, "pitchfork as a whole is experimental", I guess it's fine to leave this as is for now. We can break peoples' workflow.